feat: custom styling and content (#217)
pingdotgg/uploadthing: docs/src/pages/theming.mdx

import { Callout } from "nextra-theme-docs";

# Theming

## Components anatomy

### UploadButton

Simplified component structure:

```jsx

> <div data-state={getUtState()} className={className}>

Reviewer: What is `getUtState`? Not something users should care about

Author: Then probably makes sense to replace it with inline comment saying something like `current state` or `readying, ready or uploading`(to specify values that can be here)

  </div>
</div>
```

<Callout type="info">
  UploadButton consists of three themable elements: container, button and
  allowed content.
</Callout>

<Callout type="warning">
  Note: UploadButton's button element is defined using `label`
</Callout>

import Image from "next/image";

<Image
  src="/upload-button-anatomy.png"
  alt="Upload button anatomy"
  width="1546"
  height="998"
  style={{ marginTop: "1rem", borderRadius: "0.5rem" }}
/>

### UploadDropzone

Simplified component structure:

```jsx
<div className={className} data-state={getUtState()}>
  <svg data-ut-element="upload-icon" data-state={getUtState()}>
    ...
  </svg>
  <label data-ut-element="label" data-state={getUtState()}>
    /label content goes here/
    <input disabled={!ready} />
  </label>
  <div data-ut-element="allowed-content" data-state={getUtState()}>
    /allowed content goes here/
  </div>
  <button
    data-ut-element="button"
    data-state={getUtState()}
    disabled={isUploading}
  >
    /button content goes here/
  </button>
</div>
```

<Callout type="info">
  UploadDropzone consists of five themable elements: container, upload icon,
  label, button and allowed content.
</Callout>

> <Callout type="warning">

Reviewer: It is not quite clear to me what this means, maybe we should refactor the components to make them have the same anatomy?

Author: Regarding what that means: Visually buttons in both UploadButton and UploadDropzone are the similar. On the markup level they are defined using different tags. In case of `UploadButton` it's `label` because it's wrapping hidden file input. To make styling through data attributes(and custom data classes from TW specifically) work in the same way among both `UploadButton` and `UploadDropzone` I introduced `data-ut-element` data attribute. So if `UploadButton` has some styles applied to button element, for example, `ut-button:text-cyan-500`, it can be just copy pasted to `UploadDropzone` and those styles will be applied to the button of `UploadDropzone`. Hope my explanation is clear in this case

Regarding anatomy refactoring: I guess we kind of can do that. We either need to replace `label` with `button` inside of `UploadButton` or `button` with `label` inside of `UploadDropzone`. Maybe there is also another option I'm missing. What do you think?

  `data-ut-element="button"`.
</Callout>

<Image
  src="/upload-dropzone-anatomy.png"
  alt="Upload dropzone anatomy"
  width="2010"
  height="1594"
  style={{ marginTop: "1rem", borderRadius: "0.5rem" }}
/>

## Theming props

### `className`

Both UploadButton and UploadDropzone accept `className` prop. It allows you to
pass any additional classes to the component. All classes that are being passed
through this prop are going to be applied to the outermost element - container

### `appearance`

Both `UploadButton` and `UploadDropzone` accept `appearance` prop. It accepts an
object with keys that correspond to elements of component. You can pass
following types of values:

- `string` - it will be applied as a class to the element
- `CSSProperties` object - it will be applied as an inline style to the element
- `callback` - it will be called with current state of the component and should
  return one of the above types

The interfaces of `appearance` prop for `UploadButton` and `UploadDropzone` are
following

```ts
type UploadButtonProps = {
  /* rest of props */
  appearance?: {
    container?: StyleField;
    button?: StyleField;
    allowedContent?: StyleField;
  };

> };

Reviewer: Might be clearer to have `interface StyleField {}` and some short inline comments

Author: Okay, will add

type UploadDropzoneProps = {
  /* rest of props */
  appearance?: {
    container?: StyleField;
    uploadIcon?: StyleField;
    label?: StyleField;
    allowedContent?: StyleField;
    button?: StyleField;
  };
};
```

The types of callback arguments are following:

```ts

> type ButtonCallbackArguments = {

Reviewer: Do we really need this callback? Is the data attributes not enough?

Author: While data attributes are enough when it comes to styling through classes, callback makes it possible to apply state-specific styles with inline styles

  ready: boolean;
  isUploading: boolean;
  uploadProgress: number;
  fileTypes: string[];
};

type DropzoneCallbackArguments = {
  ready: boolean;
  isUploading: boolean;
  uploadProgress: number;
  fileTypes: string[];
  isDragActive: boolean;
};
```

## How to theme

> ### With TailwindCSS

Reviewer: Might make sense to use `<Tabs>` here. There's already quite a lot of text and scrolling necessary to get where you wanna be.

Author: I intentionally didn't use tabs to make use of headings that builds up navigation on the right. It makes it possible to quickly find what you need, click on it and page will scroll to the heading. 

Do you think it's an overkill in this case and tabs should be enough?

#### Configuring TailwindCSS

To leverage best developer experience, we strongly recommend extending your
TailwindCSS config using our utility function `withUt`

> <Callout type="warning">

Reviewer: Could be rephrased a bit. Don't think it needs to be a warning either

Author: How would you rephrase it? Give me at least some direction

Reviewer 2: I would probably pull it out of the callout, and add it to the above paragraph as something like:

> This utility function adds additional classes and variants used to style the uploadthing components. If you choose not to use it, be sure to import the default stylesheet into your app.

  Note: `withUt` is optional but if you do not use it, you will not be able to
  use new classes and variants it adds. Also, in case you are not using it, make
  sure to import default stylesheet into your app.
</Callout>

```ts filename="tailwind.config.ts"
import { withUt } from "@uploadthing/react/tw";

/** @type {import('tailwindcss').Config} */
export default withUt({
  // your config goes here
});
```

```js filename="tailwind.config.js"
// @ts-check
const { withUt } = require("@uploadthing/react/tw");

/** @type {import('tailwindcss').Config} */
module.exports = withUt({
  // your config goes here
});

> ```

Reviewer: A language picker in top right would be nice instead of showing both, although out of scope for this PR

Author: Doesn't look like it's something available in Nextra out of the box. If you know open source project that implemented such thing in a proper way, I could make use of their code ;D

> <Callout type="warning">

Reviewer: This shouldn't be necessary if we rephrase the one above I think

  Note: Once you extended your config, it is recommended to remove default
  stylesheet import
</Callout>

`withUt` does two major things:

- Adds new entry to `content` property so that it includes UploadThing's
  components. It allows to avoid having duplicated styles in your bundle.
- Registers plugin and adds nice-to-have classes to improve styling experience.

#### `className` prop

`className` prop accepts any classes so you can pass there anything you like.
When it comes to TailwindCSS, to make styling easier, `withUt` adds few utility
classes and variants for you to leverage.

Tailwind variants that allow applying styles to internal elements and their
arbitrary variant alternatives:

> | Custom variant       |                        Arbitrary variant |

Reviewer: Why should users care about this table?

Author: I thought it might be useful for particular users to know what arbitrary variants can be used instead of custom classes and variants. Though thinking about it now, I guess components anatomy should be enough to understand how to target children of component's container

| :------------------- | ---------------------------------------: |
| `ut-button`          |          `&>*[data-ut-element="button"]` |
| `ut-allowed-content` | `&>*[data-ut-element="allowed-content"]` |
| `ut-label`           |           `&>*[data-ut-element="label"]` |
| `ut-upload-icon`     |     `&>*[data-ut-element="upload-icon"]` |

Tailwind classes that allow applying styles based on state of element and their
arbitrary `data` class alternatives:

| Custom class   |          Arbitrary class |
| :------------- | -----------------------: |
| `ut-readying`  |  `data-[state=readying]` |
| `ut-ready`     |     `data-[state=ready]` |
| `ut-uploading` | `data-[state=uploading]` |

These variants and classes can be used in conjunction with each other to make
component match your design in the exact way you want.

```jsx
<UploadButton
    className="bg-blue-500 ut-button:bg-cyan-500 ut-button:ut-readying:bg-cyan-500/50"
               ^           ^ applied to button   ^ applied to button when loading config
               | applied to container

> />

Reviewer: Very nice. We could also render this component beneath to show the result

Author: Makes sense, will do

Author: Would screenshots be enough or you want to render the component within the markdown?

Reviewer: > Would screenshots be enough or you want to render the component within the markdown?

Adding @uploadthing/react as a dep and just render it would probably be the best.

Author: @juliusmarminge But doesn't FE component require API endpoint to be present to pull configuration from? I mean I'm not sure If I can define API endpoint in Nextra

Reviewer: Nextra is just Nextjs, so you could.

<UploadDropzone
    className="bg-blue-500 ut-label:text-lg ut-allowed-content:ut-uploading:text-red-300"
               ^           ^                ^ applied to allowed content when uploading file
               |           | applied to label
               | applied to container

/>
```

#### `appearance` prop

If you need, you can pass TailwindCSS classes directly to specific elements of
component or provide a callback that will be called with current state of the
component and will return a string

```jsx
<UploadButton
  appearance={{

>     button({ ready, isUploading }) {

Reviewer: Why should I use the callback here instead of tailwind variants?

Author: Well, it would definitely make more sense for a user to leverage TW variants instead of callback for state-specific styling. I guess I can remove usage of callback in this particular example to avoid promotion of not the best practices

    uploadIcon?: StyleField;
    label?: StyleField;
    allowedContent?: StyleField;
    button?: StyleField;
  };
};
```

where the `CallbackArguments` is defined as (depending on the component):

```ts
type ButtonCallbackArguments = {
  ready: boolean;
  isUploading: boolean;
  uploadProgress: number;
  fileTypes: string[];
};

type DropzoneCallbackArguments = {
  ready: boolean;
  isUploading: boolean;
  uploadProgress: number;
  fileTypes: string[];
  isDragActive: boolean;
};
```

## How to theme

### With TailwindCSS

#### Configuring TailwindCSS

To leverage the best developer experience, we strongly recommend wrapping your
Tailwind config with our utility function `withUt`. This utility function adds
additional classes and variants used to style our components.

In addition, it also automatically sets the `content` option to include all the
necessary classes that the components uses. This allows you to avoid having
duplicated styles in your bundle. Therefore, when using `withUt`, you should not
import our stylesheet into your app. If you choose not to use `withUt`, you have
to import default stylesheet to make components look right.

```ts filename="tailwind.config.ts"

> import { withUt } from "@uploadthing/react/tw";

Reviewer: wrong import

/** @type {import('tailwindcss').Config} */
export default withUt({
  // your config goes here
});
```

```js filename="tailwind.config.js"
// @ts-check

> const { withUt } = require("@uploadthing/react/tw");

Reviewer: same as above

import { Callout } from "nextra-theme-docs";

import { UploadButton, UploadDropzone } from "@uploadthing/react";

> # Theming

Reviewer 3: Only thought here - it'd be nice to have a minimal example up top and a bit more context. Not your responsibility, will put it on my TODO

### UploadDropzone

Simplified component structure:

```jsx
<div className={className} data-state={/* "ready" | "readying" | "uploading" */}>
  <svg data-ut-element="upload-icon" data-state={/* "ready" | "readying" | "uploading" */}>
    ...
  </svg>
  <label data-ut-element="label" data-state={/* "ready" | "readying" | "uploading" */}>
    {/* label content goes here */}
    <input disabled={!ready} />
  </label>
  <div data-ut-element="allowed-content" data-state={/* "ready" | "readying" | "uploading" */}>
    /allowed content goes here/
  </div>
  <button
    data-ut-element="button"
    data-state={/* "ready" | "readying" | "uploading" */}
    disabled={isUploading}
  >
    {/* button content goes here */}
  </button>
</div>
```

<Callout type="info">
  UploadDropzone consists of five themable elements: `container`, `upload icon`,
  `label`, `button` and `allowed content`.
</Callout>

<Callout type="warning">
  Note: While in UploadButton the button element is defined using `label`, in
  UploadDropzone it is defined using `button`. As an abstraction layer, button
  element in these two components have special data attribute applied:
  `data-ut-element="button"`.
</Callout>

> <Image

Reviewer 3: These diagrams are so good

```ts
type ButtonCallbackArguments = {
  ready: boolean;
  isUploading: boolean;
  uploadProgress: number;
  fileTypes: string[];
};

type DropzoneCallbackArguments = {
  ready: boolean;
  isUploading: boolean;
  uploadProgress: number;
  fileTypes: string[];
  isDragActive: boolean;
};
```

## How to theme

### With TailwindCSS

#### Configuring TailwindCSS

To leverage the best developer experience, we strongly recommend wrapping your
Tailwind config with our utility function `withUt`. This utility function adds
additional classes and variants used to style our components.

In addition, it also automatically sets the `content` option to include all the
necessary classes that the components uses. This allows you to avoid having
duplicated styles in your bundle. Therefore, when using `withUt`, you should not
import our stylesheet into your app. If you choose not to use `withUt`, you have
to import default stylesheet to make components look right.

```ts filename="tailwind.config.ts"
import { withUt } from "uploadthing/tw";

export default withUt({
  // your config goes here
});
```

```js filename="tailwind.config.js"
// @ts-check

> const { withUt } = require("uploadthing/tw");

Reviewer 3: Surprisingly elegant solution

import { Callout } from "nextra-theme-docs";
import { twMerge } from "tailwind-merge";

# Theming

## Components anatomy

### UploadButton

Simplified component structure:

```jsx
<div data-state={/* "ready" | "readying" | "uploading" */} className={className}>
  <label data-state={/* "ready" | "readying" | "uploading" */} data-ut-element="button">
    <input />
    /button content goes here/
  </label>
  <div data-state={/* "ready" | "readying" | "uploading" */} data-ut-element="allowed-content">
    /allowed content text goes here/
  </div>
</div>
```

<Callout type="info">

>   UploadButton consists of three themable elements: container, button and
>   allowed content.

Reviewer: ```suggestion
  UploadDropzone consists of five themable elements: `container`, `upload icon`,
  `label`, `button` and `allowed content`.
```

</Callout>

<Callout type="warning">
  Note: UploadButton's button element is defined using `label`
</Callout>

import Image from "next/image";

<Image
  src="/upload-button-anatomy.png"
  alt="Upload button anatomy"
  width="1546"
  height="998"
  style={{ marginTop: "1rem", borderRadius: "0.5rem" }}
/>

### UploadDropzone

Simplified component structure:

```jsx
<div className={className} data-state={/* "ready" | "readying" | "uploading" */}>
  <svg data-ut-element="upload-icon" data-state={/* "ready" | "readying" | "uploading" */}>
    ...
  </svg>
  <label data-ut-element="label" data-state={/* "ready" | "readying" | "uploading" */}>
    /label content goes here/
    <input disabled={!ready} />
  </label>
  <div data-ut-element="allowed-content" data-state={/* "ready" | "readying" | "uploading" */}>
    /allowed content goes here/
  </div>
  <button
    data-ut-element="button"
    data-state={/* "ready" | "readying" | "uploading" */}
    disabled={isUploading}
  >
    /button content goes here/
  </button>
</div>
```

<Callout type="info">

>   UploadDropzone consists of five themable elements: container, upload icon,
>   label, button and allowed content.

Reviewer: ```suggestion
  UploadDropzone consists of five themable elements: `container`, `upload icon`,
  `label`, `button` and `allowed content`.
```

</Callout>

<Callout type="warning">
  Note: While in UploadButton the button element is defined using `label`, in
  UploadDropzone it is defined using `button`. As an abstraction layer, button
  element in these two components have special data attribute applied:
  `data-ut-element="button"`.
</Callout>

<Image
  src="/upload-dropzone-anatomy.png"
  alt="Upload dropzone anatomy"
  width="2010"
  height="1594"
  style={{ marginTop: "1rem", borderRadius: "0.5rem" }}
/>

## Theming props

### `className`

> Both UploadButton and UploadDropzone accept `className` prop. It allows you to
> pass any additional classes to the component. All classes that are being passed
> through this prop are going to be applied to the outermost element - container

Reviewer: ```suggestion
Both UploadButton and UploadDropzone accept `className` prop. It allows you to
pass any additional classes to the component. All classes that are being passed
through this prop are going to be applied to the outermost element - container.
```

| `ut-uploading:`       | Applied to the container element when the component is uploading. |

</div>

<Callout>
  If you're not using the `withUt` wrapper, the state variants can be applied
  using `data-[state="..."]:`
</Callout>

These variants and classes can be used in conjunction with each other to make
component match your design in the exact way you want.

```tsx
<UploadButton
    className="mt-6 ut-button:bg-red-500 ut-button:ut-readying:bg-red-500/50"
               |    └─ applied to `button`  └─ applied to `button` when `readying`
               └─ applied to `container`
/>
```

<div
  className={twMerge(
    "flex flex-col items-center justify-center gap-1",
    "ut-button:bg-red-500 ut-button:ut-readying:bg-red-500/50 mt-6",
  )}
>
  <label
    className="relative flex h-10 w-36 cursor-pointer items-center justify-center overflow-hidden rounded-md bg-blue-600 text-white after:transition-[width] after:duration-500"
    data-ut-element="button"
  >
    <input className="sr-only" disabled />
    <p>Ready</p>
  </label>
  <div
    className="h-[1.25rem] text-xs leading-5 text-white"
    data-ut-element="allowed-content"
  >
    <p>Allowed content</p>
  </div>

> </div>

Reviewer: Can we not setup a mock endpoint & router and use the components from the react lib? Will make it more future proof

Author: How to handle demonstration of state-specific styles in this case? I mean `readying` state is available only during quite short amount of time after component mount. State `uploading` requires actually uploading files and also visible for short amount of time

Reviewer: hmm - we could add a `__internal-state` prop or something to control the state (doesn't even have to be exposed as a type) and use that?

```mdx
<UploadButton __internal_state="ready" endpoint="foobar" />
<UploadButton __internal_state="readying" endpoint="foobar" />
<UploadButton __internal_state="uploading" endpoint="foobar" />
```

Author: I was thinking about something like that as well. I also think that if that state is used we actually should make components non interactive. I mean that we don't actually want users be able to upload anything. Also we don't need to actually fetch config and therefore we don't need exposing an API route. What do you think?

```jsx
<UploadButton
  appearance={{
    button({ ready, isUploading }) {
      return `${ready ? "bg-green-500" : "bg-red-500"} ${
        isUploading ? "cursor-not-allowed" : ""
      } after:bg-orange-400`;

>     },

Reviewer: i didn't mean remove this part, just migrate it to use the tailwind variants

```ts
    button: "after:bg-orange-400 bg-red-500 ut-ready:bg-green-500 ut-uploading:cursor-not-allowed"
```
